File: src/effects/AudioReverser.cpp
#include "AudioReverser.h"
#include <algorithm>

AudioReverser::AudioReverser() {
}

AudioReverser::~AudioReverser() {
}

AudioBuffer AudioReverser::reverse(const AudioBuffer& input) {
    // 최적화: reverse iterator로 직접 생성 (복사 1회로 감소)
    const std::vector<float>& inputData = input.getData();
    std::vector<float> data(inputData.rbegin(), inputData.rend());

    // 결과 버퍼 생성
    AudioBuffer result(input.getSampleRate(), input.getChannels());
    result.setData(std::move(data)); // move semantics 사용

    return result;
}

File: web/js/js/effects/AudioReverser.js
import { AudioBuffer } from '../audio/AudioBuffer.js';

/**
 * AudioReverser - 오디오 역재생 (JavaScript 버전)
 */
export class AudioReverser {
    constructor() {}

    /**
     * 오디오를 역재생
     * @param {AudioBuffer} input
     * @returns {AudioBuffer}
     */
    reverse(input) {
        const output = new AudioBuffer(input.getSampleRate(), input.getChannels());
        const data = input.getData();
        const reversed = new Float32Array(data.length);

        // 샘플을 역순으로 배치
        for (let i = 0; i < data.length; i++) {
            reversed[i] = data[data.length - 1 - i];
        }

        output.setData(reversed);
        return output;
    }
}
